### PR TITLE
Feat: Fix memory issue by replacing io.ReadAll with io.Copy (#4017)

### LINKED ISSUE #4017: KServe Agent OOMKilled When Allocated Memory is Insufficient
/kind bug

**What steps did you take and what happened:**
I attempted to load a model using the KServe Agent with the `StorageURI` pointing to **Google Cloud Storage**. However, when the memory allocated to the KServe Agent was smaller than the memory required by the model, the agent was OOMKilled and subsequently restarted.

**What did you expect to happen:**
I expected the KServe Agent to attempt downloading the model using minimal memory resources, regardless of the model's size. This approach would allow models of varying sizes to be loaded without being restricted by the initially allocated memory, enhancing flexibility and reliability in different deployment scenarios.

**What's the InferenceService yaml:**
KServe/agent OOMKilled

**Anything else you would like to add**
I plan to address this issue and will be submitting a Pull Request (PR) to fix it. I would appreciate it if the maintainers could review my PR once it's submitted.

**Environment:**

Same Kubeflow Manifest
- Istio Version: 1.22.1
- Knative Version: 1.12.4
- KServe Version: 0.13.1
- Kubeflow version: 1.9.0
- Cloud Environment: On-prem, GKE
- Kubernetes version: v1.30.3
- OS: Ubuntu 20.04

### PR DESCRIPTION
**Pull request by ops-jaeha**
**What this PR does / why we need it**:
Fixes an out-of-memory (OOM) issue that occurs when downloading large files from Google Cloud Storage (GCS). 
Replaces `io.ReadAll()` with `io.Copy()` to stream data and prevent excessive memory usage.

**Which issue(s) this PR fixes** :
Fixes #4017

**Type of changes**
- [x] Bug fix (non-breaking change which fixes an issue)

**Feature/Issue validation/testing**:
Test A : Replaced `io.ReadAll()` with `io.Copy()` in the GCS download function.
Test B : Tested downloading large files from GCS to ensure memory usage remains stable.
Test C : Verified that existing functionality is unaffected by the change.

- Logs
```
{"severity":"INFO","timestamp":"2024-10-25T05:59:19.932967054Z","caller":"logging/config.go:80","message":"Unable to read vcs.revision from binary"}
{"severity":"INFO","timestamp":"2024-10-25T05:59:19.933186058Z","caller":"agent/main.go:146","message":"Initializing model agent with config-dir /mnt/configs, model-dir /mnt/models"}
{"severity":"INFO","timestamp":"2024-10-25T05:59:19.933205723Z","caller":"agent/syncer.go:37","message":"Syncing from model dir /mnt/models"}
{"severity":"INFO","timestamp":"2024-10-25T05:59:19.933291351Z","caller":"agent/main.go:311","message":"Starting puller"}
{"severity":"INFO","timestamp":"2024-10-25T05:59:19.933308112Z","caller":"agent/main.go:161","message":"Starting agent http server..."}
{"severity":"INFO","timestamp":"2024-10-25T05:59:19.933419621Z","caller":"agent/watcher.go:95","message":"Start to watch model config event"}
{"severity":"INFO","timestamp":"2024-10-25T05:59:19.933462608Z","caller":"agent/main.go:336","message":"Building server user port 8080 port 9081"}
{"severity":"INFO","timestamp":"2024-10-25T05:59:19.933486366Z","caller":"agent/watcher.go:133","message":"Watching /mnt/data/models.json"}
{"severity":"INFO","timestamp":"2024-10-25T06:00:27.561580441Z","caller":"agent/watcher.go:109","message":"Processing event CREATE        \"/mnt/configs/..data\""}
{"severity":"INFO","timestamp":"2024-10-25T06:00:27.561973675Z","caller":"agent/watcher.go:180","message":"adding model my-model"}
{"severity":"INFO","timestamp":"2024-10-25T06:00:27.562043687Z","caller":"agent/puller.go:144","message":"Worker is started for my-model"}
{"severity":"INFO","timestamp":"2024-10-25T06:00:27.562067421Z","caller":"agent/puller.go:152","message":"Downloading model from gs://my-bucket/models/my-model"}
{"severity":"INFO","timestamp":"2024-10-25T06:00:27.562145489Z","caller":"agent/downloader.go:47","message":"Downloading gs://my-bucket/models/my-model to model dir /mnt/models"}
{"severity":"INFO","timestamp":"2024-10-25T06:00:27.567088747Z","logger":"modelAgent","caller":"storage/gcs.go:39","message":"Downloading model ","modelName":"my-model","storageUri":"gs://my-bucket/models/my-model","modelDir":"/mnt/models"}
{"severity":"INFO","timestamp":"2024-10-25T06:00:48.268679916Z","logger":"modelAgent","caller":"storage/gcs.go:157","message":"Wrote  to file /mnt/models/my-model/1/my-model.safetensors"}
{"severity":"INFO","timestamp":"2024-10-25T06:00:48.31997657Z","logger":"modelAgent","caller":"storage/gcs.go:157","message":"Wrote  to file /mnt/models/my-model/config.pbtxt"}
{"severity":"INFO","timestamp":"2024-10-25T06:00:48.320341167Z","caller":"agent/downloader.go:73","message":"Creating successFile /mnt/models/my-model/SUCCESS.c355724640c567783469f4df19fa01faffea8df1a7c422910c722d9a83e95b2f"}
{"severity":"INFO","timestamp":"2024-10-25T06:00:56.840635114Z","caller":"agent/puller.go:177","message":"Successfully loaded model my-model"}
{"severity":"INFO","timestamp":"2024-10-25T06:00:56.84074305Z","caller":"agent/puller.go:137","message":"completion event for model my-model, in flight ops 0"}

```

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Fix out-of-memory issue when downloading large files from GCS by replacing io.ReadAll with io.Copy to stream data and optimize memory usage.

### CODE REVIEW

File: pkg/agent/storage/gcs.go [hunk 1]
```diff
@@ -131,33 +131,16 @@ func (g *GCSObjectDownloader) DownloadFile(client stiface.Client, attrs *gstorag
...
+	
+	if _, err := io.Copy(file, reader); err != nil {
+        return fmt.Errorf("failed to copy object(%s) from bucket(%s) to file: %w",
+            attrs.Name,
+            attrs.Bucket,
+            err,
+        )
+    }
 
-func (g *GCSObjectDownloader) WriteToFile(data []byte, attrs *gstorage.ObjectAttrs, file *os.File) error {
```
Comment (spolti) [inline pkg/agent/storage/gcs.go:145]: Can't this func be deleted?

**Comment by spolti** [comment]
Hi @ops-jaeha, thanks for the fix. Can you please sign your commit?

File: pkg/agent/storage/gcs.go (hunk 1 above)
Comment (ops-jaeha) [reply pkg/agent/storage/gcs.go:145]: I don't understand about your comment.
I'm already delete WriteToFile function.

Comment (spolti) [reply pkg/agent/storage/gcs.go:145]: Yes, correct, the way the deletions and additions was structured has confused me :D

**Review by spolti — APPROVED** [review]

**Comment by ops-jaeha** [comment]
I think it is a signed commit, is that correct?
I'm not very familiar with signed commits since it's my first time.

**Comment by ops-jaeha** [comment]
> Hi @ops-jaeha, thanks for the fix. Can you please sign your commit?

I think it is a signed commit, is that correct?
I'm not very familiar with signed commits since it's my first time.

**Comment by spolti** [comment]
when pushing the commit you need to sign it with the `-s` option.
As the PR is already created you can update it by using `git commit -s --amend`

**Comment by ops-jaeha** [comment]
I have made a signed commit and confirmed that it has been verified. What should I do next to proceed with CI?

**Comment by spolti** [comment]
> I have made a signed commit and confirmed that it has been verified. What should I do next to proceed with CI?
Did you push?

**Comment by ops-jaeha** [comment]
> > I have made a signed commit and confirmed that it has been verified. What should I do next to proceed with CI?

> Did you push?

Yes, I think so, but I'll try again.

**Comment by ops-jaeha** [comment]
I'm finished DCO

**Comment by ops-jaeha** [comment]
@spolti Can you approve workflows?

**Comment by spolti** [comment]
@yuzisun @terrytangyuan can you please help with this?

**Comment by sivanantha321** [comment]
> @spolti Can you approve workflows?

approved

**Comment by ops-jaeha** [comment]
Hello @spolti @sivanantha321 
Due to a recent system change on my end, it seems that I need re-approval for this PR. I apologize for the extra step, but could I kindly ask for your approval again?

Also, I’m curious: when the workflow fails, will I need to request another approval every time I make a commit to address the issues, or is it retained?

Thank you for your patience!

**Comment by sivanantha321** [comment]
> Hello @spolti @sivanantha321 Due to a recent system change on my end, it seems that I need re-approval for this PR. I apologize for the extra step, but could I kindly ask for your approval again?
> 
> Also, I’m curious: when the workflow fails, will I need to request another approval every time I make a commit to address the issues, or is it retained?
> 
> Thank you for your patience!

Yeah. It needs approval every time you make a push. This is only applicable for the first time contribution. Also can you help sign off the commits ? The DCO check is failing.

**Comment by ops-jaeha** [comment]
> > Hello @spolti @sivanantha321 Due to a recent system change on my end, it seems that I need re-approval for this PR. I apologize for the extra step, but could I kindly ask for your approval again?
> > Also, I’m curious: when the workflow fails, will I need to request another approval every time I make a commit to address the issues, or is it retained?
> > Thank you for your patience!
> 
> Yeah. It needs approval every time you make a push. This is only applicable for the first time contribution. Also can you help sign off the commits ? The DCO check is failing.

I'll try again.

**Comment by ops-jaeha** [comment]
> > > Hello @spolti @sivanantha321 Due to a recent system change on my end, it seems that I need re-approval for this PR. I apologize for the extra step, but could I kindly ask for your approval again?
> > > Also, I’m curious: when the workflow fails, will I need to request another approval every time I make a commit to address the issues, or is it retained?
> > > Thank you for your patience!
> > 
> > 
> > Yeah. It needs approval every time you make a push. This is only applicable for the first time contribution. Also can you help sign off the commits ? The DCO check is failing.
> 
> I'll try again.

I’ve completed up to the DCO.

**Comment by ops-jaeha** [comment]
@spolti @sivanantha321 
All workflows have successfully passed.

**Comment by ops-jaeha** [comment]
This was my first time contributing to a large open-source project, and thanks to @spolti and @sivanantha321, 
I was able to experience this for the first time. I truly appreciate your kindness and support.

**Comment by ops-jaeha** [comment]
You also need to rebuild and push the KServe Agent Docker file. Please handle that.

**Comment by yuzisun** [comment]
> You also need to rebuild and push the KServe Agent Docker file. Please handle that.

Should be handled by the KServe CI, thanks.

**Comment by sivanantha321** [comment]
> Should be handled by the KServe CI, thanks.

I just noticed that workflows are not triggered if it is automerged by the github-actions-bot.

**Pull request merged**